### PR TITLE
CORE-11922 - Fix issue with MemberListCache

### PR DESCRIPTION
**Pull request by nikinagy**
**Description of the issue:**
MemberListCache replaces entries based on name check. This means that we are not able to store penguin and non-pending information in it at the same time.

**Fix:**
Do a status check: 
- if new entry has status active or suspended, then replace old entry with same name having non-pending status
- if new entry has pending status, then replace old entry with same name and status

**Testing:**
To test these changes I had to modify the lookup API to return not just the active/suspended, but pending information too.

Registered MGM and two members, Alice and Bob.

**MGM calling the lookup API:**

_**Picture:**_
mgm-lookup

_**Result as string:**_
`
{"members":[
{"memberContext":{"corda.cpi.name":"cpi name","corda.cpi.signer.summary.hash":"SHA-256:367DDC08BB0BFBC8B338E2B8DC17EB1715A542386E6FE2376A9FB9EBC80A3DEC","corda.cpi.version":"1.0.0.0-SNAPSHOT","corda.ecdh.key":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEBu+OEejHWDl3jKpMEc1AXnqUoncy\noTUJi/5QFnewLmGoJzWox+hk6nGWLGmXjGLlr7s3SOqGxP7Vjk0gc0R5cQ==\n-----END PUBLIC KEY-----\n","corda.endpoints.0.connectionURL":"https://corda-p2p-gateway-worker.corda-cluster-a:8080","corda.endpoints.0.protocolVersion":"1","corda.groupId":"da0966f1-de87-411a-8444-2d1259df0181","corda.name":"O=MGM, L=London, C=GB","corda.platformVersion":"5000","corda.session.keys.0.hash":"1CACF6945553AE3688F1B838E543551625ACE9E55A4ECACFD7474C112B7299A7","corda.session.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE5380Y0mskY/JD8FHeXhgP79BgcPW\nmuriM7ZI04dQyeHWIioQlsK14Fp6CAD0DVcTlZTnqiTFhXJHYNkCO/0aFQ==\n-----END PUBLIC KEY-----\n","corda.softwareVersion":"5.0.0.0-SNAPSHOT"},"mgmContext":{"corda.creationTime":"2023-03-24T19:56:49.320938Z","corda.mgm":"true","corda.modifiedTime":"2023-03-24T19:56:49.320938Z","corda.serial":"1","corda.status":"ACTIVE"}},
{"memberContext":{"corda.cpi.name":"cpi name2","corda.cpi.signer.summary.hash":"SHA-256:367DDC08BB0BFBC8B338E2B8DC17EB1715A542386E6FE2376A9FB9EBC80A3DEC","corda.cpi.version":"1.0.0.0-SNAPSHOT","corda.endpoints.0.connectionURL":"https://corda-p2p-gateway-worker.corda-cluster-a:8080","corda.endpoints.0.protocolVersion":"1","corda.groupId":"da0966f1-de87-411a-8444-2d1259df0181","corda.ledger.keys.0.hash":"001B3BB61212859F8F5A1932488FBCE35ED8726A36EE4B4EFE9176B0F6312063","corda.ledger.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEYJcEj4CqetHLCiPRdiZa45s8WzwS\n1CSO2/fz+45Ky/jocQI76VgZqgzQZQZ4breVGNwjyW4nKyRnio0J+rGCpA==\n-----END PUBLIC KEY-----\n","corda.ledger.keys.0.signature.spec":"SHA256withECDSA","corda.name":"O=Alice, L=London, C=GB","corda.platformVersion":"5000","corda.registration.id":"d8761dd6-a837-47f2-a4c6-9c0ec20178c4","corda.session.keys.0.hash":"653944F0B62D1F469C59993B6CB227BCBFDADFE592BFEB3B8E28A5CFAF8B5C3A","corda.session.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAED8vm/FAhqU1goI0rSUyWl+vlMf/n\nIhUXCeBIYDoNguEg3jDMzKte522gZeUfrD+EalJL4j+KhfnLM5Jn5iN6kw==\n-----END PUBLIC KEY-----\n","corda.session.keys.0.signature.spec":"SHA256withECDSA","corda.softwareVersion":"5.0.0.0-SNAPSHOT"},"mgmContext":{"corda.creationTime":"2023-03-24T20:02:22.037630Z","corda.modifiedTime":"2023-03-24T20:02:22.037630Z","corda.serial":"1","corda.status":"ACTIVE"}},
{"memberContext":{"corda.cpi.name":"cpi name2","corda.cpi.signer.summary.hash":"SHA-256:367DDC08BB0BFBC8B338E2B8DC17EB1715A542386E6FE2376A9FB9EBC80A3DEC","corda.cpi.version":"1.0.0.0-SNAPSHOT","corda.endpoints.0.connectionURL":"https://corda-p2p-gateway-worker.corda-cluster-a:8080","corda.endpoints.0.protocolVersion":"1","corda.groupId":"da0966f1-de87-411a-8444-2d1259df0181","corda.ledger.keys.0.hash":"B0E95B9E844AD21866CDF55E0E5D0465C4CB71E366D3D6934033DDACA8B6A003","corda.ledger.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEU+gsCyxHsezWhb+OsSIqbrehak9a\njk/VnO6Md4Q2OcdZtwfO7pNCwU0D67uwPs6NHj+VumlRSavjSoziaOAHHA==\n-----END PUBLIC KEY-----\n","corda.ledger.keys.0.signature.spec":"SHA256withECDSA","corda.name":"O=Bob, L=London, C=GB","corda.platformVersion":"5000","corda.registration.id":"7664b919-e23e-44ac-b667-3a074a4d9816","corda.session.keys.0.hash":"DBD18FDF5CB5F2C86676D267CF41B7069DF05341B02766000867128EA3F688C0","corda.session.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEJuYma5Gjh8Mi2GStY7e0w14YQzaF\n6dknTls2aUoyZEXFoc9hbIzngfn3Ps4ZfqcfCub0tsBkr0iSEUOpGDeFSA==\n-----END PUBLIC KEY-----\n","corda.session.keys.0.signature.spec":"SHA256withECDSA","corda.softwareVersion":"5.0.0.0-SNAPSHOT"},"mgmContext":{"corda.creationTime":"2023-03-24T20:04:21.879139Z","corda.modifiedTime":"2023-03-24T20:04:21.879139Z","corda.serial":"1","corda.status":"ACTIVE"}},
{"memberContext":{"corda.cpi.name":"cpi name2","corda.cpi.signer.summary.hash":"SHA-256:367DDC08BB0BFBC8B338E2B8DC17EB1715A542386E6FE2376A9FB9EBC80A3DEC","corda.cpi.version":"1.0.0.0-SNAPSHOT","corda.endpoints.0.connectionURL":"https://corda-p2p-gateway-worker.corda-cluster-a:8080","corda.endpoints.0.protocolVersion":"1","corda.groupId":"da0966f1-de87-411a-8444-2d1259df0181","corda.ledger.keys.0.hash":"001B3BB61212859F8F5A1932488FBCE35ED8726A36EE4B4EFE9176B0F6312063","corda.ledger.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEYJcEj4CqetHLCiPRdiZa45s8WzwS\n1CSO2/fz+45Ky/jocQI76VgZqgzQZQZ4breVGNwjyW4nKyRnio0J+rGCpA==\n-----END PUBLIC KEY-----\n","corda.ledger.keys.0.signature.spec":"SHA256withECDSA","corda.name":"O=Alice, L=London, C=GB","corda.platformVersion":"5000","corda.registration.id":"d8761dd6-a837-47f2-a4c6-9c0ec20178c4","corda.session.keys.0.hash":"653944F0B62D1F469C59993B6CB227BCBFDADFE592BFEB3B8E28A5CFAF8B5C3A","corda.session.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAED8vm/FAhqU1goI0rSUyWl+vlMf/n\nIhUXCeBIYDoNguEg3jDMzKte522gZeUfrD+EalJL4j+KhfnLM5Jn5iN6kw==\n-----END PUBLIC KEY-----\n","corda.session.keys.0.signature.spec":"SHA256withECDSA","corda.softwareVersion":"5.0.0.0-SNAPSHOT"},"mgmContext":{"corda.creationTime":"2023-03-24T20:02:22.037630Z","corda.modifiedTime":"2023-03-24T20:02:22.037630Z","corda.serial":"1","corda.status":"PENDING"}},
{"memberContext":{"corda.cpi.name":"cpi name2","corda.cpi.signer.summary.hash":"SHA-256:367DDC08BB0BFBC8B338E2B8DC17EB1715A542386E6FE2376A9FB9EBC80A3DEC","corda.cpi.version":"1.0.0.0-SNAPSHOT","corda.endpoints.0.connectionURL":"https://corda-p2p-gateway-worker.corda-cluster-a:8080","corda.endpoints.0.protocolVersion":"1","corda.groupId":"da0966f1-de87-411a-8444-2d1259df0181","corda.ledger.keys.0.hash":"B0E95B9E844AD21866CDF55E0E5D0465C4CB71E366D3D6934033DDACA8B6A003","corda.ledger.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEU+gsCyxHsezWhb+OsSIqbrehak9a\njk/VnO6Md4Q2OcdZtwfO7pNCwU0D67uwPs6NHj+VumlRSavjSoziaOAHHA==\n-----END PUBLIC KEY-----\n","corda.ledger.keys.0.signature.spec":"SHA256withECDSA","corda.name":"O=Bob, L=London, C=GB","corda.platformVersion":"5000","corda.registration.id":"7664b919-e23e-44ac-b667-3a074a4d9816","corda.session.keys.0.hash":"DBD18FDF5CB5F2C86676D267CF41B7069DF05341B02766000867128EA3F688C0","corda.session.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEJuYma5Gjh8Mi2GStY7e0w14YQzaF\n6dknTls2aUoyZEXFoc9hbIzngfn3Ps4ZfqcfCub0tsBkr0iSEUOpGDeFSA==\n-----END PUBLIC KEY-----\n","corda.session.keys.0.signature.spec":"SHA256withECDSA","corda.softwareVersion":"5.0.0.0-SNAPSHOT"},"mgmContext":{"corda.creationTime":"2023-03-24T20:04:21.879139Z","corda.modifiedTime":"2023-03-24T20:04:21.879139Z","corda.serial":"1","corda.status":"PENDING"}}
]}%  
`

**Bob calling the lookup API:**

_**Picture:**_
bob-lookup

_**Result as string:**_
`
{"members":[
{"memberContext":{"corda.cpi.name":"cpi name","corda.cpi.signer.summary.hash":"SHA-256:367DDC08BB0BFBC8B338E2B8DC17EB1715A542386E6FE2376A9FB9EBC80A3DEC","corda.cpi.version":"1.0.0.0-SNAPSHOT","corda.ecdh.key":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEBu+OEejHWDl3jKpMEc1AXnqUoncy\noTUJi/5QFnewLmGoJzWox+hk6nGWLGmXjGLlr7s3SOqGxP7Vjk0gc0R5cQ==\n-----END PUBLIC KEY-----\n","corda.endpoints.0.connectionURL":"https://corda-p2p-gateway-worker.corda-cluster-a:8080","corda.endpoints.0.protocolVersion":"1","corda.groupId":"da0966f1-de87-411a-8444-2d1259df0181","corda.name":"O=MGM, L=London, C=GB","corda.platformVersion":"5000","corda.session.keys.0.hash":"1CACF6945553AE3688F1B838E543551625ACE9E55A4ECACFD7474C112B7299A7","corda.session.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE5380Y0mskY/JD8FHeXhgP79BgcPW\nmuriM7ZI04dQyeHWIioQlsK14Fp6CAD0DVcTlZTnqiTFhXJHYNkCO/0aFQ==\n-----END PUBLIC KEY-----\n","corda.softwareVersion":"5.0.0.0-SNAPSHOT"},"mgmContext":{"corda.creationTime":"2023-03-24T20:02:53.580752Z","corda.mgm":"true","corda.modifiedTime":"2023-03-24T20:02:53.580752Z","corda.serial":"1","corda.status":"ACTIVE"}},
{"memberContext":{"corda.cpi.name":"cpi name2","corda.cpi.signer.summary.hash":"SHA-256:367DDC08BB0BFBC8B338E2B8DC17EB1715A542386E6FE2376A9FB9EBC80A3DEC","corda.cpi.version":"1.0.0.0-SNAPSHOT","corda.endpoints.0.connectionURL":"https://corda-p2p-gateway-worker.corda-cluster-a:8080","corda.endpoints.0.protocolVersion":"1","corda.groupId":"da0966f1-de87-411a-8444-2d1259df0181","corda.ledger.keys.0.hash":"B0E95B9E844AD21866CDF55E0E5D0465C4CB71E366D3D6934033DDACA8B6A003","corda.ledger.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEU+gsCyxHsezWhb+OsSIqbrehak9a\njk/VnO6Md4Q2OcdZtwfO7pNCwU0D67uwPs6NHj+VumlRSavjSoziaOAHHA==\n-----END PUBLIC KEY-----\n","corda.ledger.keys.0.signature.spec":"SHA256withECDSA","corda.name":"O=Bob, L=London, C=GB","corda.platformVersion":"5000","corda.registration.id":"7664b919-e23e-44ac-b667-3a074a4d9816","corda.session.keys.0.hash":"DBD18FDF5CB5F2C86676D267CF41B7069DF05341B02766000867128EA3F688C0","corda.session.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEJuYma5Gjh8Mi2GStY7e0w14YQzaF\n6dknTls2aUoyZEXFoc9hbIzngfn3Ps4ZfqcfCub0tsBkr0iSEUOpGDeFSA==\n-----END PUBLIC KEY-----\n","corda.session.keys.0.signature.spec":"SHA256withECDSA","corda.softwareVersion":"5.0.0.0-SNAPSHOT"},"mgmContext":{"corda.creationTime":"2023-03-24T20:04:21.879139Z","corda.modifiedTime":"2023-03-24T20:04:21.879139Z","corda.serial":"1","corda.status":"ACTIVE"}},
{"memberContext":{"corda.cpi.name":"cpi name2","corda.cpi.signer.summary.hash":"SHA-256:367DDC08BB0BFBC8B338E2B8DC17EB1715A542386E6FE2376A9FB9EBC80A3DEC","corda.cpi.version":"1.0.0.0-SNAPSHOT","corda.endpoints.0.connectionURL":"https://corda-p2p-gateway-worker.corda-cluster-a:8080","corda.endpoints.0.protocolVersion":"1","corda.groupId":"da0966f1-de87-411a-8444-2d1259df0181","corda.ledger.keys.0.hash":"001B3BB61212859F8F5A1932488FBCE35ED8726A36EE4B4EFE9176B0F6312063","corda.ledger.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEYJcEj4CqetHLCiPRdiZa45s8WzwS\n1CSO2/fz+45Ky/jocQI76VgZqgzQZQZ4breVGNwjyW4nKyRnio0J+rGCpA==\n-----END PUBLIC KEY-----\n","corda.ledger.keys.0.signature.spec":"SHA256withECDSA","corda.name":"O=Alice, L=London, C=GB","corda.platformVersion":"5000","corda.registration.id":"d8761dd6-a837-47f2-a4c6-9c0ec20178c4","corda.session.keys.0.hash":"653944F0B62D1F469C59993B6CB227BCBFDADFE592BFEB3B8E28A5CFAF8B5C3A","corda.session.keys.0.pem":"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAED8vm/FAhqU1goI0rSUyWl+vlMf/n\nIhUXCeBIYDoNguEg3jDMzKte522gZeUfrD+EalJL4j+KhfnLM5Jn5iN6kw==\n-----END PUBLIC KEY-----\n","corda.session.keys.0.signature.spec":"SHA256withECDSA","corda.softwareVersion":"5.0.0.0-SNAPSHOT"},"mgmContext":{"corda.creationTime":"2023-03-24T20:02:22.037630Z","corda.modifiedTime":"2023-03-24T20:02:22.037630Z","corda.serial":"1","corda.status":"ACTIVE"}}
]}%  
`

### CODE REVIEW

File: components/membership/membership-group-read-impl/src/main/kotlin/net/corda/membership/impl/read/cache/MemberListCache.kt [hunk 1]
```diff
@@ -36,7 +40,12 @@ interface MemberListCache : MemberDataListCache<MemberInfo> {
             cache.compute(holdingIdentity) { _, value ->
                 (value ?: ReplaceableList())
                     .addOrReplace(data) { old, new ->
-                        old.name == new.name
+                        if(new.status == MEMBER_STATUS_ACTIVE || new.status == MEMBER_STATUS_SUSPENDED) {
```
Comment (dimosr) [inline components/membership/membership-group-read-impl/src/main/kotlin/net/corda/membership/impl/read/cache/MemberListCache.kt]: I hadn't reviewed the original code, so might be missing some context. What was the reasoning behind the use of `ReplaceableList`? Is there a reason for example, we didn't use a hash map and on lookup just retrieve all the values and convert them to a list if that's the collection we wanted to expose?

It seems like what we are doing via `addOrReplace` here is some form of implicit keying? That becomes even more implicit now that the logic becomes more complicated (and for me personally hard to follow) . Should we just key per (member name, isPending) the same way we do at the topic level? I think it would be consistent and easier to understand.

(Note: guava contains a [`Table`](https://guava.dev/releases/19.0/api/docs/com/google/common/collect/Table.html) collection that helps you do that quite easily)

Comment (nikinagy) [reply components/membership/membership-group-read-impl/src/main/kotlin/net/corda/membership/impl/read/cache/MemberListCache.kt]: @charlieR3 might be the best person to answer this, because I don't remember why we went with ReplaceableList

Comment (charlieR3) [reply components/membership/membership-group-read-impl/src/main/kotlin/net/corda/membership/impl/read/cache/MemberListCache.kt]: It's not fresh in my memory now, but we went through a number of iterations on this and took into account performance of look ups and potential pitfalls in thread safety.

For reference, here is the original PR and you can see in the comments the various approaches tried. https://github.com/corda/corda-runtime-os/pull/565#issuecomment-1009999848

That link links to the comment where we landed on `ReplaceableList` as a solution with good performance for look up in comparison to other approaches while also providing thread safety. 

If I remember right, the main issue was if a service reads the member list, and then while iterating it, the underlying list changed so the service iterating the list ran in to an exception. The replaceable list just creates a new list so any service currently holding a reference won't run into an exception when iterating the list they have a reference to.

Comment (dimosr) [reply components/membership/membership-group-read-impl/src/main/kotlin/net/corda/membership/impl/read/cache/MemberListCache.kt]: It feels like we are potentially optimising something that might not even be a bottleneck at the moment (which is also very likely given membership is not expected to be a high-traffic component) at the cost of code readability/maintainability. 

> If I remember right, the main issue was if a service reads the member list, and then while iterating it, the underlying list changed so the service iterating the list ran in to an exception. The replaceable list just creates a new list so any service currently holding a reference won't run into an exception when iterating the list they have a reference to.

I think what I was suggesting above would not suffer from that, given the `getAll()` operation could still get all the values and return them in a new list, which could also be immutable if preferred. Any thoughts on this?

Comment (charlieR3) [reply components/membership/membership-group-read-impl/src/main/kotlin/net/corda/membership/impl/read/cache/MemberListCache.kt]: > It feels like we are potentially optimising something that might not even be a bottleneck at the moment (which is also very likely given membership is not expected to be a high-traffic component) at the cost of code readability/maintainability.

It's quite possibly we may have tried to optimise this before there was a bottleneck, yes. However, we always envisioned that while in general the membership components will not be high traffic, the read operations on membership lookup _would_ be high traffic which was why we went through that exercise of trying to find what we thought was the best option at the time. 

What you describe sounds like it would work also, although we would need to create a new list on every read whereas the current solution would only create a new list on writes which we expect to be a much less frequent operation. 

I'm not opposed to changing it as you've said for cleaner and more readable code. Could be worth doing a quick sanity check around the performance differences if making that change to check for no major impact. I suspect it would only really show a performance difference in a high traffic network. The test used previously for that is still checked in as disabled (MemberListCachePerformanceTest).

Comment (dimosr) [reply components/membership/membership-group-read-impl/src/main/kotlin/net/corda/membership/impl/read/cache/MemberListCache.kt]: That said, it doesn't need to hold this up I think. Worth hearing what you think anyway and we could schedule something for later, if needed.

If we were to keep the `ReplaceableList` for now, would it be worth at least swapping around the logic so that it's a bit more tolerant in cases where we add new statuses (a bit on the fence about it myself)? Something like the below, for example:
```
if (new.status == MEMBER_STATUS_PENDING) {
    old.status == MEMBER_STATUS_PENDING && old.name == new.name
} else {
    old.status != MEMBER_STATUS_PENDING && old.name == new.name
}
```

Comment (nikinagy) [reply components/membership/membership-group-read-impl/src/main/kotlin/net/corda/membership/impl/read/cache/MemberListCache.kt]: This new status check really looks better and more readable than mine, so happy to modify it.

Comment (dimosr) [reply components/membership/membership-group-read-impl/src/main/kotlin/net/corda/membership/impl/read/cache/MemberListCache.kt]: ok, let's stay with what we have for now then. We might have to revisit this as part of the large network support work anyway, so not worth spending too much time on it at the moment.

**Review by dimosr — APPROVED** [review]
LGTM
